### ISSUE TITLE
Add an auto-cleare bloom filter

### ISSUE THREAD

**Issue by doubiliu** [neo-project/neo#2606]
**Summary or problem description**

When node receives a transaction, it needs to check whether the transaction exists to avoid repeated transactions.The current implementation logic is by querying in levelDB,which caused plenty of query operations.Of course, we used the Bloom filter function of levelDB. However, according to the characteristics of the Bloom filter, adding a large amount of data inside it will increase the possibility of judgment errors.Therefore, levelDB still needs to read data from disk.

Therefore, we can add a Bloom filter that can be automatically cleared as a pre-judgment before querying levelDB to reduce the possibility of misjudgment and reduce IO reading from disk.

**Do you have any solution you want to propose?**
This function requires two new Bloom filters built in.

**Process**
1. From the start of the node, all transactions in [0h,24h) will be added to the first bloom filter.
2. All transactions in [24h,48h] will be added to the second bloom filter. 
3. [48h ,72h) will be added to the first Bloom filter (cleared first). 
Then it will cycle in turn.

When we check a transaction, we need first check whether the first bloom filter exists, and then check whether the second bloom filter exists. If both of them are determined to be non-existent, it can be considered as not a duplicate transaction. If it is determined that it may exist, then perform the original check operation.Don't worry about omissions, the reason is that the transaction validity period is 24 hours, so even if there are duplicate transactions after cleaning up, it will still be judged as a failure.The only check that may be needed is that within the first 48 hours, it still needs to be read from levelDB, because there are some periodic interval calculation problems．

This can reduce  IO reading from disk.

**Neo Version**
- Neo 3

**Where in the software does this update applies to?**
- Ledger

**Comment by Qiao-Jin** [comment]
@erikzhang @shargon @roman-khimov Could you please have a look?

**Comment by roman-khimov** [comment]
Let me show you two pictures:
<img width="650" alt="Screenshot_20210922_124844" src="https://user-images.githubusercontent.com/22092804/134322230-f706a8d7-b885-4b92-88ff-63387461dd7c.png">
<img width="659" alt="Screenshot_20210922_124954" src="https://user-images.githubusercontent.com/22092804/134322243-28fe9e08-ef00-4de4-999b-aa9c3c065f20.png">

The second one performs `HasTransaction` check somewhat like five times faster, so it may seem that it should work much better. But real TPS difference is about 2-3%.

We have [multiple DB backends](https://github.com/nspcc-dev/neo-go/blob/master/docs/node-configuration.md#db-configuration) in neo-go with quite different behavior (but nspcc-dev/neo-go#2130). And we have [neo-bench](https://github.com/nspcc-dev/neo-bench/) that we use to test things. The pictures above are from runs with LevelDB and BoltDB, but if I'm to run recent neo-go with `inmemory` backend (that only became possible recently) it still won't give more than 5% TPS improvement over Level/Bolt. Of course that depends on machine, lower number of cores and slower disks can make this difference bigger.

But even though LevelDB really has some cost of key (non-)existence check, I'm not sure it's worth optimizing it.

**Comment by roman-khimov** [comment]
Speaking of `inmemory` or MemoryStore, btw. Now that we can test it, there are some other pictures to share:
![tps_single_10_wrk](https://user-images.githubusercontent.com/22092804/134337116-3fa30cfb-7783-4624-ac96-a0f9ef040c5a.png)
![tpb_single_10_wrk](https://user-images.githubusercontent.com/22092804/134337125-b89a49f5-978c-47fd-ad0a-101de516baba.png)
![ms_per_block_single_10_wrk](https://user-images.githubusercontent.com/22092804/134337130-b9546c59-0010-4f60-b6d2-8bdab37cdfb1.png)

Aside from MemoryStore scalability issues (it obviously wasn't _intended_ for processing of 1M of transactions, so I can't blame it) I'd note that initial blocks (where the store is almost empty) show about the same metrics as with LevelDB, although transaction (non)existence check should be much faster with MemoryStore.

**Comment by doubiliu** [comment]
Maybe it has something to do with the test sample size. @Qiao-Jin also conducted some tests before, but there are still some differences. @Qiao-Jin Can you give me a link?

**Comment by Qiao-Jin** [comment]
Yes, I have tested this benchmark before in another issue, where significant TPS differences of TPS exists between 3 scenarios:
(1) Basic logic, where new TXs have to be checked in levelDB
(2) LevenDB with self-attached bloom filter applied (leveldb_filterpolicy_create_bloom)
(3) Tx cache employed so that don't need to check levelDB

The third scenario is different from this issue. However, effect should be similiar as in this issue, it should be sufficient to check most new-coming Txs by bloomfilter only. 

https://github.com/neo-project/neo/issues/1734#issuecomment-651035576

I think the optimization effect upon levelDB checking is related to DB size. The bigger DB size is, the more time will be used to look up a key, and optimization effect of similiar logic will be more significant. In my case over 5M Txs are used and I think optimization effect can be better if DB size is bigger.

**Comment by Liaojinghui** [comment]
@roman-khimov Really love your benchmark result. May I know the name of the tool that you use to conduct it?

**Comment by roman-khimov** [comment]
> @roman-khimov Really love your benchmark result. May I know the name of the tool that you use to conduct it?

...........:arrow_down: 
:arrow_right: [neo-bench](https://github.com/nspcc-dev/neo-bench/) :arrow_left: 
...........:arrow_up: 
(`plot.py` might need to be tweaked depending on what you test and want to visualize)

I've not tried 5M transactions with it, although it can be changed to do that too. In any event, if you think the proposed change is helpful be my guest and implement it (it's obviously not a month effort to make a prototype), as long as it doesn't change the protocol I can easily test it and tell you the result, maybe there is something to it.

**Comment by Liaojinghui** [comment]
@roman-khimov awesome

**Comment by roman-khimov** [comment]
I can elaborate a little on C# node performance if you'd like. That's opinionated, of course, but there are reasons why I'm somewhat skeptical about these optimizations. We've spent some time optimizing neo-go and it basically implements the same protocol, so the key problems are the same.

So a node is roughly a transaction processing machine and we want to make it process as many transactions as we can, if we're to follow transaction route it can be simplified as:
1. Verifying and accepting transaction into mempool
2. Taking it from mempool, packing into a block along with other transactions
3. Executing transaction along with other ones in a block

That's quite a natural pipeline, but with a loop:
 * stages 1 and 2 are synchronized via mempool
 * stage 3 is directly dependent on 2
 * stage 3 then affects mempool again (kicking out processed transactions)
 * at the same time stages 1 and 2-3 can work in parallel, the node can accept new transactions while creating/processing block

Stage 1 is:
 * dominated by crypto and tx (non)existence check
 * can be done in parallel for multiple transactions up to the point of entrance to the mempool

Stage 2 is not very interesting, its mostly about signing a block.

Stage 3 is:
 * dominated by contract execution
 * strictly sequential for all transactions included in a block

What we can see from MemoryStore/LevelDBStore comparison above is that there is no real difference in initial phase of the test, even though underlying DB is radically different and this should influence tx existence check and contract execution (we need to get/put things into the DB anyway).

The first six blocks for LevelDBStore look like this:
```
DeltaTime, TransactionsCount, TPS
1027, 10939, 10651.412
1076, 7871, 7315.056
1043, 9653, 9255.034
1047, 7829, 7477.555
1061, 8912, 8399.623
1059, 8592, 8113.314
```
For MemoryStore:
```
DeltaTime, TransactionsCount, TPS
1015, 10641, 10483.744
1048, 8186, 7811.069
1030, 9506, 9229.126
1041, 8071, 7753.122
1040, 8908, 8565.385
1039, 8443, 8126.083
```

When the test starts it pushes as many transactions as it can into the empty mempool (stage 1), then the timer triggers in one second and takes everything it can from the mempool to create a block (stage 2). As we can see only 10-11K transactions get into the mempool during this first second of the test irrespective of the DB backend (even though LevelDB is almost empty, it will still give slightly higher overhead for tx check).

Then we can also see that LevelDBStore TPS metrics are very close to straight line during whole test run with block times being very close to the target one second, so the node processes the block, removes old transactions from the mempool and when the time comes takes another batch of transactions from it, this batch is always ~8K transactions.

Which suggests that the first bottleneck is transaction verification/adding into the mempool (stage 1), block times are very stable during whole test run so we can conclude that one second is enough for stage 3 processing of the same 8K transactions. And (surprise!) the change proposed tries to optimize stage 1, so I shouldn't be skeptical about it, but I'm.

And that's because I know that 8K or 10K TPS is _very_ low for stage 1. Let's try being optimistic and expect a 50% improvement from this change. That would be 12-15K transactions for the first block. Sorry, but that still is _very_ low. As mentioned above this stage is dominated by crypto and DB check and in general ECDSA verification takes more time than DB check (especially for initial blocks). I assume we can do ~10K ECDSA verifications per second per any decent core (please check, if not your crypto library is not up to the task) and stage 1 can be done in parallel for the most part. So why do we have just 10K transactions in the mempool for the first block? I don't know, I only have a rough understanding of how C# node works, but I'd suggest that it's just not using multiple cores efficiently for stage 1.

That to me seems to be the biggest problem right now, so what I would do is:
 * forget about profiler, don't look at it (it'll be useful, but later)
 * forget about crypto, assume it works fine
 * forget about LevelDB, assume it works fine
 * trace exact tx path from the source (P2P or RPC) to mempool, make sure that the only synchronous part of that route is addition to the mempool (of already verified not-yet-seen transaction), take a look at #2107 also just in case
 * the first target here is 20-25K transactions per second added to the mempool (neo-go easily accepts 50K in one second and I'm 100% sure C# node can do the same, there is no magic there, we do the same crypto and we use the same DB)
 * after that most probably you will have some problems with stage 3 and that's where profiler might be handy (given that you still ignore crypto/DB)

I might be wrong of course, but that's my thoughts for now. After solving this and adapting block processing to the increased number of transactions we might also move on to multinode and multiaddress testing scenarios, testing GAS/voting NEO/non-native NEP-17 transfers (which are all possible now with neo-bench), there are lots of interesting scenarios, but we need to deal with the basic one first.

**Comment by roman-khimov** [comment]
BTW (sorry, this topic is too addictive), I've tried neo-bench with 5M transactions and it looks interesting. I don't see any big difference for C# node (10 and 30 are number of threads pushing transactions) which I think proves my previous point:
![tps_single](https://user-images.githubusercontent.com/22092804/134555620-0db2302c-6c5b-4d0e-8502-0cb7f0a96237.png)

But for neo-go this LevelDB deterioration becomes much more visible:
![blb2](https://user-images.githubusercontent.com/22092804/134556372-4b6796f6-4529-4aaa-b61e-da7d6db74589.png)

Luckily we have BoltDB and it easily beats LevelDB in this test (with 1M transactions they're pretty close), but that's a bit different story.

**Comment by doubiliu** [comment]
Our analysis and guess is related to the storage structure and query mode of levelDB. Since levelDB is a tree-shaped multi-layer file structure, if we want to determine that some data does not exist, the database may need to traverse many data files. And if it perform the merge operation, the data will be disrupted, which is the reason why it consumes a lot of disk IO.

**Comment by roman-khimov** [comment]
Sure, no doubt about it. I'm not saying it's wrong, LevelDB is LevelDB and the way its structured obviously affects its operation. My point is that probably it's not the main problem for C# node. But I can be wrong also, so if you implement this and it will make C# node perform twice faster --- I'm all for it. I'm just trying to share some experiences of ours here and provide some alternative view, but in the end of the day anything that helps C# node performance is good (as long as we don't change the protocol, doing that needs some very specific justification).

**Comment by doubiliu** [comment]
Thank you very much for the analysis, which helped us better understand the current problems. I think neo’s performance still has space for improvement. Let’s think about whether there is a better way together.
@roman-khimov
